lib/SPIRVProducerPass.cpp

```diff
@@ -2439,8 +2439,8 @@ void SPIRVProducerPass::GenerateGlobalVar(GlobalVariable &GV) {
       HasMD = HasMDVec[0];
       for (uint32_t i = 1; i < HasMDVec.size(); i++) {
         if (HasMD != HasMDVec[i]) {
-          llvm_unreachable(
-              "Kernels should have consistent work group size definition");
+          HasMD = false;
```

Comment:
If I've understood the logic correctly, I think this can all be simplified. We generate spec constants if there are any kernels that don't use `reqd_work_group_size`, or if there are conflicting values. You don't need this loop or a `bool` vector, you just need a single `bool` to track whether you can use a constant wgsize, initialize it to `true`, and set it to `false` when you encounter a kernel that hasn't got `reqd_work_group_size` or if there's a conflict.

Reply:
Conflicting values are ok. If you take a look at some of the tests we can generate `LocalSize` execution mode. It's only when some kernels don't have it we need to have the spec constants (and no execution mode). I'll double check the logic here.

Reply:
I think this part of the code is only hit if the kernels use the work-group size though (e.g. they call `get_local_size()`), and if they do (and they're using different `reqd_work_group_size`) then we don't generate `LocalSize` execution mode. Maybe worth adding a test case for that scenario.

I'm not disagreeing with the logic that's here - I think it's correct - I'm just saying the same thing can be written without the vector and the loop.

Reply:
We add the global variable (which is what this deals with) if any kernel does not use required workgroup size **or** if one of the builtins (e.g. get_local_size()) that requires it is used.

In any event, I'll double check this logic.

Reply:
I've removed the loop.